# Captone Project Submission

Please fill out:
* Student name: Wesley Yu
* Student pace: Flex
* Scheduled project review date/time: May 03, 2023 / 3:30 PM
* Instructor name: Abhineet Kulkarni
* Blog post URL:https://github.com/whyyoutoo/Sentiment_Classification_for_Amazon_Product_Reviews


# Merging and Subsetting Data

Grabbing asins that contain heaphone or earphone in their title.

In [1]:
# importing nesseccary libraries
import pandas as pd
import json
import gzip
import numpy as np

In [2]:
# read in data to pandas dataframe as per instructions
def parse(path):
  g = gzip.open(path, 'rb')
  for l in g:
    yield json.loads(l)

def getDF(path):
  i = 0
  df = {}
  for d in parse(path):
    df[i] = d
    i += 1
  return pd.DataFrame.from_dict(df, orient='index')

df = getDF('data/Electronics_5.json.gz')

In [3]:
# read in data to pandas dataframe
def parse(path):
  g = gzip.open(path, 'rb')
  for l in g:
    yield json.loads(l)

def getDF(path):
  i = 0
  df = {}
  for d in parse(path):
    df[i] = d
    i += 1
  return pd.DataFrame.from_dict(df, orient='index')

df2 = getDF('data/meta_Electronics.json.gz')

## Product Meta Data

In [4]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 786445 entries, 0 to 786444
Data columns (total 19 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   category         786445 non-null  object
 1   tech1            786445 non-null  object
 2   description      786445 non-null  object
 3   fit              786445 non-null  object
 4   title            786445 non-null  object
 5   also_buy         786445 non-null  object
 6   tech2            786445 non-null  object
 7   brand            786445 non-null  object
 8   feature          786445 non-null  object
 9   rank             786445 non-null  object
 10  also_view        786445 non-null  object
 11  main_cat         786445 non-null  object
 12  similar_item     786445 non-null  object
 13  date             786445 non-null  object
 14  price            786445 non-null  object
 15  asin             786445 non-null  object
 16  imageURL         786445 non-null  object
 17  imageURLHi

Meta data provides information of 786,445 different products

In [5]:
df2.head()

category tech1  \
0  [Electronics, Camera &amp; Photo, Video Survei...         
1                  [Electronics, Camera &amp; Photo]         
2  [Electronics, eBook Readers &amp; Accessories,...         
3  [Electronics, eBook Readers & Accessories, eBo...         
4  [Electronics, eBook Readers & Accessories, eBo...         

                                         description fit  \
0  [The following camera brands and models have b...       
1  [This second edition of the Handbook of Astron...       
2  [A zesty tale. (Publishers Weekly)<br /><br />...       
3                                                 []       
4  [&#8220;sex.lies.murder.fame. is brillllli&#82...       

                                               title  \
0  Genuine Geovision 1 Channel 3rd Party NVR IP S...   
1  Books "Handbook of Astronomical Image Processi...   
2                                     One Hot Summer   
3  Hurray for Hattie Rabbit: Story and pictures (...   
4                     sex.lies.murder.fame.: A Novel   

                               also_buy tech2  \
0                                    []         
1                          [0999470906]         
2              [0425167798, 039914157X]         
3  [0060219521, 0060219580, 0060219394]         
4                                    []         

                                          brand  \
0                                     GeoVision   
1                                  33 Books Co.   
2  Visit Amazon's Carolina Garcia Aguilera Page   
3           Visit Amazon's Dick Gackenbach Page   
4              Visit Amazon's Lolita Files Page   

                                             feature  \
0  [Genuine Geovision 1 Channel NVR IP Software, ...   
1  [Detailed chapters cover these fundamental top...   
2                                                 []   
3                                                 []   
4                                                 []   

                                                rank  \
0  [>#3,092 in Tools &amp; Home Improvement &gt; ...   
1  [>#55,933 in Camera &amp; Photo (See Top 100 i...   
2                               3,105,177 in Books (   
3                               2,024,298 in Books (   
4                               3,778,828 in Books (   

                              also_view            main_cat similar_item  \
0                                    []  Camera &amp; Photo                
1  [0943396670, 1138055360, 0999470906]  Camera &amp; Photo                
2                                    []               Books                
3  [0060219521, 0060219475, 0060219394]               Books                
4                                    []               Books                

               date                                              price  \
0  January 28, 2014                                             $65.00   
1     June 17, 2003                                                      
2                                                               $11.49   
3                    .a-section.a-spacing-mini{margin-bottom:6px!im...   
4                                                               $13.95   

         asin                                           imageURL  \
0  0011300000  [https://images-na.ssl-images-amazon.com/image...   
1  0043396828  [https://images-na.ssl-images-amazon.com/image...   
2  0060009810                                                 []   
3  0060219602                                                 []   
4  0060786817                                                 []   

                                     imageURLHighRes details  
0  [https://images-na.ssl-images-amazon.com/image...     NaN  
1  [https://images-na.ssl-images-amazon.com/image...     NaN  
2                                                 []     NaN  
3                                                 []     NaN  
4                                                 []     NaN

I am only interested in headphone or earphone products, first I will convert title feature to lowercase so i can filter more easily.

In [6]:
# convert listing names to lowercase
df2.title = df2.title.apply(lambda x: x.lower())

Subsetting meta data to container products with only headphone or earphone in product title.

In [7]:
# filter products with only heaphone or earpone in listing
product_df = df2[df2.title.str.contains('headphone|headphone|earphone|earphones')]

## Review Data

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6739590 entries, 0 to 6739589
Data columns (total 12 columns):
 #   Column          Dtype  
---  ------          -----  
 0   overall         float64
 1   vote            object 
 2   verified        bool   
 3   reviewTime      object 
 4   reviewerID      object 
 5   asin            object 
 6   style           object 
 7   reviewerName    object 
 8   reviewText      object 
 9   summary         object 
 10  unixReviewTime  int64  
 11  image           object 
dtypes: bool(1), float64(1), int64(1), object(9)
memory usage: 623.5+ MB


5-core electronics data contains over 6 million reviews.

In [9]:
df.head()

overall vote  verified   reviewTime      reviewerID        asin  \
0      5.0   67      True  09 18, 1999   AAP7PPBU72QFM  0151004714   
1      3.0    5      True  10 23, 2013  A2E168DTVGE6SV  0151004714   
2      5.0    4     False   09 2, 2008  A1ER5AYS3FQ9O3  0151004714   
3      5.0   13     False   09 4, 2000  A1T17LMQABMBN5  0151004714   
4      3.0    8      True   02 4, 2000  A3QHJ0FXK33OBE  0151004714   

                            style      reviewerName  \
0       {'Format:': ' Hardcover'}      D. C. Carrad   
1  {'Format:': ' Kindle Edition'}               Evy   
2       {'Format:': ' Paperback'}             Kcorn   
3       {'Format:': ' Hardcover'}   Caf Girl Writes   
4       {'Format:': ' Hardcover'}  W. Shane Schmidt   

                                          reviewText  \
0  This is the best novel I have read in 2 or 3 y...   
1  Pages and pages of introspection, in the style...   
2  This is the kind of novel to read when you hav...   
3  What gorgeous language! What an incredible wri...   
4  I was taken in by reviews that compared this b...   

                                             summary  unixReviewTime image  
0                                     A star is born       937612800   NaN  
1                    A stream of consciousness novel      1382486400   NaN  
2  I'm a huge fan of the author and this one did ...      1220313600   NaN  
3          The most beautiful book I have ever read!       968025600   NaN  
4                        A dissenting view--In part.       949622400   NaN

Merging meta data and reviews data with inner join to keep reviews only containing headphones or earphones.

In [10]:
merged = df.merge(product_df, on = 'asin')

In [9]:
merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 471228 entries, 0 to 471227
Data columns (total 30 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   overall          471228 non-null  float64
 1   vote             65135 non-null   object 
 2   verified         471228 non-null  bool   
 3   reviewTime       471228 non-null  object 
 4   reviewerID       471228 non-null  object 
 5   asin             471228 non-null  object 
 6   style            294015 non-null  object 
 7   reviewerName     471157 non-null  object 
 8   reviewText       471168 non-null  object 
 9   summary          471148 non-null  object 
 10  unixReviewTime   471228 non-null  int64  
 11  image            9026 non-null    object 
 12  category         471228 non-null  object 
 13  tech1            24660 non-null   object 
 14  description      471228 non-null  object 
 15  fit              22 non-null      object 
 16  title            471228 non-null  obje

Total 471,228 reviews remain. Exploration of features and null values will be conducted in EDA notebook.

In [11]:
# saving subsetted dataset.
merged.to_csv('data/headphone_reviews.csv', index = False)